### PR TITLE
feat: add raw block volume support

### LINKED ISSUE #27: [feature] Implement raw block volume support
https://kubernetes-csi.github.io/docs/raw-block.html

### PR DESCRIPTION
**Pull request by ZeroMagic**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
add raw block volume support

**Which issue(s) this PR fixes**:

Fixes #27 

**Special notes for your reviewer**:
We can use the yaml files in `deploy/example/rawblock` to test the feature manully, and verify it according to the README.md.

**Release note**:
```
feat: add raw block volume support
```

### CODE REVIEW

File: pkg/azuredisk/nodeserver.go [hunk 1]
```diff
@@ -351,3 +395,44 @@ func getFStype(attributes map[string]string) string {
 
 	return ""
 }
+
+func (d *Driver) findDiskAndLun(devicePath string) (string, int32, error) {
```
Comment (andyzhangx) [inline pkg/azuredisk/nodeserver.go]: add comments and example for this function

File: pkg/azuredisk/nodeserver.go [hunk 2]
```diff
@@ -190,6 +178,14 @@ func (d *Driver) NodePublishVolume(ctx context.Context, req *csi.NodePublishVolu
 		return nil, status.Error(codes.InvalidArgument, "Target path not provided")
 	}
 
+	isBlock := req.GetVolumeCapability().GetBlock() != nil
```
Comment (andyzhangx) [inline pkg/azuredisk/nodeserver.go]: may have issue here, looks like there is another way:
```
	switch mode := volCap.GetAccessType().(type) {
	case *csi.VolumeCapability_Block:
		if err := d.nodePublishVolumeForBlock(req, mountOptions); err != nil {
			return nil, err
		}
	case *csi.VolumeCapability_Mount:
		if err := d.nodePublishVolumeForFileSystem(req, mountOptions, mode); err != nil {
			return nil, err
		}
	}
```

Comment (ZeroMagic) [reply pkg/azuredisk/nodeserver.go]: done.

File: pkg/azuredisk/nodeserver.go (hunk 1 above)
Comment (ZeroMagic) [reply pkg/azuredisk/nodeserver.go]: done.

**Comment by andyzhangx** [comment]
/hold
I will cut off a v0.3.0 beta release for this driver first and then merge this PR.

File: pkg/azuredisk/nodeserver.go [hunk 3]
```diff
@@ -351,3 +395,48 @@ func getFStype(attributes map[string]string) string {
 
 	return ""
 }
+
+// findDiskAndLun: returns the new device absolute path and LUN(Logical Unit Number).
+// Before node stage and publish, the devicePath has been handle to a LUN number.
+// So in this function, devicePath is always a LUN number.
+// eg. devicePath is 1, returns /sys/bus/scsi/devices/3:0:0:1 and 1.
```
[the remark this answers is not in the document]
Comment (ZeroMagic) [reply pkg/azuredisk/nodeserver.go]: done.

**Comment by andyzhangx** [comment]
@ZeroMagic do we have any test to cover this? sanity test, e2e test?

**Comment by ZeroMagic** [comment]
Later I will write an e2e test.

File: pkg/azuredisk/nodeserver.go [hunk 4]
```diff
@@ -341,6 +333,58 @@ func (d *Driver) NodeExpandVolume(ctx context.Context, req *csi.NodeExpandVolume
 	return nil, status.Error(codes.Unimplemented, fmt.Sprintf("NodeExpandVolume is not yet implemented"))
 }
 
+func (d *Driver) publishBlockVolume(req *csi.NodePublishVolumeRequest) error {
```
Comment (andyzhangx) [inline pkg/azuredisk/nodeserver.go]: is it possible to combine this `publishBlockVolume` with original `NodePublishVolume` func? there are too many common code

Comment (ZeroMagic) [reply pkg/azuredisk/nodeserver.go]: done.

**Comment by ZeroMagic** [comment]
The new e2e test `should create a raw block volume on demand` is passed.
After combining the common code, all e2e tests have been passed.

```
• [SLOW TEST:112.887 seconds]
Dynamic Provisioning
/root/go/src/github.com/kubernetes-sigs/azuredisk-csi-driver/test/e2e/dynamic_provisioning.go:32
  should create a volume on demand and mount it as readOnly in a pod
  /root/go/src/github.com/kubernetes-sigs/azuredisk-csi-driver/test/e2e/dynamic_provisioning.go:101
------------------------------
• [SLOW TEST:76.615 seconds]
[azuredisk-csi-e2e] [single-az] Pre-Provisioned
/root/go/src/github.com/kubernetes-sigs/azuredisk-csi-driver/test/e2e/pre_provisioning.go:41
  [env] should use a pre-provisioned volume and mount it as readOnly in a pod
  /root/go/src/github.com/kubernetes-sigs/azuredisk-csi-driver/test/e2e/pre_provisioning.go:78
------------------------------
• [SLOW TEST:126.696 seconds]
Dynamic Provisioning
/root/go/src/github.com/kubernetes-sigs/azuredisk-csi-driver/test/e2e/dynamic_provisioning.go:32
  should create a volume on demand
  /root/go/src/github.com/kubernetes-sigs/azuredisk-csi-driver/test/e2e/dynamic_provisioning.go:55
------------------------------
• [SLOW TEST:46.589 seconds]
Dynamic Provisioning
/root/go/src/github.com/kubernetes-sigs/azuredisk-csi-driver/test/e2e/dynamic_provisioning.go:32
  should delete PV with reclaimPolicy "Delete"
  /root/go/src/github.com/kubernetes-sigs/azuredisk-csi-driver/test/e2e/dynamic_provisioning.go:187
------------------------------
• [SLOW TEST:27.971 seconds]
[azuredisk-csi-e2e] [single-az] Pre-Provisioned
/root/go/src/github.com/kubernetes-sigs/azuredisk-csi-driver/test/e2e/pre_provisioning.go:41
  [env] should use a pre-provisioned volume and retain PV with reclaimPolicy "Retain"
  /root/go/src/github.com/kubernetes-sigs/azuredisk-csi-driver/test/e2e/pre_provisioning.go:112
------------------------------
• [SLOW TEST:186.809 seconds]
Dynamic Provisioning
/root/go/src/github.com/kubernetes-sigs/azuredisk-csi-driver/test/e2e/dynamic_provisioning.go:32
  should create a deployment object, write and read to it, delete the pod and write and read to it again
  /root/go/src/github.com/kubernetes-sigs/azuredisk-csi-driver/test/e2e/dynamic_provisioning.go:162
------------------------------
• [SLOW TEST:243.788 seconds]
Dynamic Provisioning
/root/go/src/github.com/kubernetes-sigs/azuredisk-csi-driver/test/e2e/dynamic_provisioning.go:32
  should create multiple PV objects, bind to PVCs and attach all to different pods on the same node
  /root/go/src/github.com/kubernetes-sigs/azuredisk-csi-driver/test/e2e/dynamic_provisioning.go:125
------------------------------
• [SLOW TEST:543.234 seconds]
Dynamic Provisioning
/root/go/src/github.com/kubernetes-sigs/azuredisk-csi-driver/test/e2e/dynamic_provisioning.go:32
  should create a raw block volume on demand
  /root/go/src/github.com/kubernetes-sigs/azuredisk-csi-driver/test/e2e/dynamic_provisioning.go:77
------------------------------
• [SLOW TEST:32.177 seconds]
Dynamic Provisioning
/root/go/src/github.com/kubernetes-sigs/azuredisk-csi-driver/test/e2e/dynamic_provisioning.go:32
  [env] should retain PV with reclaimPolicy "Retain"
  /root/go/src/github.com/kubernetes-sigs/azuredisk-csi-driver/test/e2e/dynamic_provisioning.go:203
------------------------------

Ran 9 of 9 Specs in 1396.768 seconds
SUCCESS! -- 9 Passed | 0 Failed | 0 Pending | 0 Skipped
PASS

Ginkgo ran 1 suite in 23m25.233341568s
Test Suite Passed
```

File: pkg/azuredisk/nodeserver.go [hunk 5]
```diff
@@ -351,3 +340,82 @@ func getFStype(attributes map[string]string) string {
 
 	return ""
 }
+
+func (d *Driver) checkMountPoint(target string) error {
```
Comment (andyzhangx) [inline pkg/azuredisk/nodeserver.go]: what does this function do? any comments? actually it does not only do `checkMountPoint`, also makes dir.

File: pkg/azuredisk/nodeserver.go [hunk 6]
```diff
@@ -190,61 +178,62 @@ func (d *Driver) NodePublishVolume(ctx context.Context, req *csi.NodePublishVolu
...
-	}
+		klog.V(2).Infof("NodePublishVolume [block]: find device path %s -> %s", devicePath, source)
 
-	if runtime.GOOS != "windows" {
-		// in windows, we will use mklink to mount, will MkdirAll in Mount func
-		if err := os.MkdirAll(target, 0750); err != nil {
-			klog.Errorf("azureDisk - mkdir failed on target: %s (%v)", target, err)
-			return nil, err
+		// here may be combined
+		globalMountPath := filepath.Dir(target)
```
Comment (andyzhangx) [inline pkg/azuredisk/nodeserver.go]: need to clarify why use parent dir here

File: pkg/azuredisk/nodeserver.go (hunk 5 above)
Comment (ZeroMagic) [reply pkg/azuredisk/nodeserver.go]: done.

File: pkg/azuredisk/nodeserver.go (hunk 6 above)
Comment (ZeroMagic) [reply pkg/azuredisk/nodeserver.go]: done.

File: pkg/azuredisk/nodeserver.go [hunk 7]
```diff
@@ -190,61 +178,62 @@ func (d *Driver) NodePublishVolume(ctx context.Context, req *csi.NodePublishVolu
...
-	}
+		klog.V(2).Infof("NodePublishVolume [block]: find device path %s -> %s", devicePath, source)
 
-	if runtime.GOOS != "windows" {
-		// in windows, we will use mklink to mount, will MkdirAll in Mount func
-		if err := os.MkdirAll(target, 0750); err != nil {
-			klog.Errorf("azureDisk - mkdir failed on target: %s (%v)", target, err)
-			return nil, err
+		// Since the block device target path is file, its parent directory should be ensured to be valid.
+		globalMountPath := filepath.Dir(target)
```
Comment (andyzhangx) [inline pkg/azuredisk/nodeserver.go]: better use `parentDir`

File: pkg/azuredisk/nodeserver.go [hunk 8]
```diff
@@ -351,3 +340,84 @@ func getFStype(attributes map[string]string) string {
...
+		}
+	}
+
+	return nil
+}
+
+// findDiskAndLun: devicePath is a LUN num, returns <device-path, lun>, e.g. </dev/sdx, 1>.
+// Before node stage and publish, the devicePath has been handle to a LUN number.
+// So in this function, devicePath is always a LUN number.
+// e.g. devicePath is 1, returns /sys/bus/scsi/devices/3:0:0:1 and 1.
```
Comment (andyzhangx) [inline pkg/azuredisk/nodeserver.go]: `/sys/bus/scsi/devices/3:0:0:1` --> `/dev/disk/azure/scsi1/lun1`

File: pkg/azuredisk/nodeserver.go [hunk 9]
```diff
@@ -351,3 +340,84 @@ func getFStype(attributes map[string]string) string {
 
 	return ""
 }
+
+// ensureMountPoint: ensure mount point to be valid.
+// If it is not existed, it will be created.
+func (d *Driver) ensureMountPoint(target string) error {
+	notMnt, err := d.mounter.IsLikelyNotMountPoint(target)
+	if err != nil && !os.IsNotExist(err) {
+		klog.V(2).Infof("azureDisk - cannot validate mount point for on %s, error: %v", target, err)
```
Comment (andyzhangx) [inline pkg/azuredisk/nodeserver.go]: remove this logging

File: pkg/azuredisk/nodeserver.go (hunk 7 above)
Comment (ZeroMagic) [reply pkg/azuredisk/nodeserver.go]: done.

File: pkg/azuredisk/nodeserver.go (hunk 8 above)
Comment (ZeroMagic) [reply pkg/azuredisk/nodeserver.go]: done

File: pkg/azuredisk/nodeserver.go (hunk 9 above)
Comment (ZeroMagic) [reply pkg/azuredisk/nodeserver.go]: done.

**Pull request merged**